elastic/beats
# Refactor state_* metricsets to share response from endpoint

## What does this PR do?
This PR changes how `kubernetes` module handle state_* metricsets which share same target endpoint.  The idea originates from https://github.com/elastic/beats/blob/master/x-pack/metricbeat/module/cloudfoundry/cloudfoundry.go

~~*Note*: At this point the PR stands more like a PoC with changes only applied at `state_container` and `state_pod` metricsets.~~ 

~~If we agree with this solution (and make sure that it would be applied with Agent too) we can extend it to the rest of `state_*` metricsets as well as to metricsets fetch from kubelet's endpoint (`node`, `pod`, `container`, `volume`, `system`)~~

In upcoming PR we will apply similar strategy for metricsets fetching from kubelet's endpoint (`node`, `pod`, `container`, `volume`, `system`)

## Why is it important?
To improve the performance of the module by avoid fetching same content multiple times.


## Checklist




## Author's Checklist

- [x] Does it have impact for Agent deployments too?
 Verified with different module's config blocks. See step 4 of testing notes below.

## How to test this PR locally

1. On a k8s cluster install kube_state_metrics (https://github.com/kubernetes/kube-state-metrics/tree/master/examples/standard)
2. Expose kube_state_metrics endpoint on the local machine -> `kubectl -n kube-system port-forward svc/kube-state-metrics 8081:8080  ` 
3. Start capturing traffic on kube-state-metrics' endpoint -> `tcpdump -i any -s 0 'tcp[((tcp[12:1] & 0xf0) >> 2):4] = 0x47455420'`
4. Test with different module's config blocks (what Agent does):
```
- module: kubernetes
  metricsets:
    - state_pod
  period: 10s
  hosts: ["0.0.0.0:8081"]
  add_metadata: false

- module: kubernetes
metricsets:
- state_container
period: 10s
hosts: ["0.0.0.0:8081"]
add_metadata: false
```

Verify with tcpdump's output that only one request takes place no matter how many modules/metricsets are enabled.

5. Test one combined module block:
```
- module: kubernetes
  metricsets:
    - state_pod
    - state_container
  period: 10s
  hosts: ["0.0.0.0:8081"]
  add_metadata: false
```

Verify with tcpdump's output that only one request takes place no matter how many modules/metricsets are enabled.

## Related issues

- Relates https://github.com/elastic/beats/issues/24869
